lightning/src/ln/onchain_utils.rs

```diff
@@ -1,8 +1,741 @@
+				counterparty_outp.write(writer)?;
+			},
+			PackageSolvingData::CounterpartyReceivedHTLCOutput(ref counterparty_outp) => {
+				3u8.write(writer)?;
+				counterparty_outp.write(writer)?;
+			},
+			PackageSolvingData::HolderHTLCOutput(ref holder_outp) => {
+				4u8.write(writer)?;
+				holder_outp.write(writer)?;
+			},
+			PackageSolvingData::HolderFundingOutput(ref funding_outp) => {
+				5u8.write(writer)?;
+				funding_outp.write(writer)?;
+			}
+		}
+		Ok(())
+	}
+}
+
+impl Readable for PackageSolvingData {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		let byte = <u8 as Readable>::read(reader)?;
+		let solving_data = match byte {
+			0 => {
+				PackageSolvingData::RevokedOutput(Readable::read(reader)?)
+			},
+			1 => {
+				PackageSolvingData::RevokedHTLCOutput(Readable::read(reader)?)
+			},
+			2 => {
+				PackageSolvingData::CounterpartyOfferedHTLCOutput(Readable::read(reader)?)
+			},
+			3 => {
+				PackageSolvingData::CounterpartyReceivedHTLCOutput(Readable::read(reader)?)
+			},
+			4 => {
+				PackageSolvingData::HolderHTLCOutput(Readable::read(reader)?)
+			},
+			5 => {
+				PackageSolvingData::HolderFundingOutput(Readable::read(reader)?)
+			}
+			_ => return Err(DecodeError::UnknownVersion)
+		};
+		Ok(solving_data)
+	}
+}
+
+/// A malleable package might be aggregated with other packages to save on fees.
+/// A untractable package has been counter-signed and aggregable will break cached counterparty
+/// signatures.
+#[derive(Clone, PartialEq)]
+pub(crate) enum PackageMalleability {
+	Malleable,
+	Untractable,
+}
+
+impl Writeable for PackageMalleability {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		match self {
+			PackageMalleability::Malleable => {
+				0u8.write(writer)?;
+			},
+			PackageMalleability::Untractable => {
+				1u8.write(writer)?;
+			}
+		}
+		Ok(())
+	}
+}
+
+impl Readable for PackageMalleability {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		let byte = <u8 as Readable>::read(reader)?;
+		let malleability = match byte {
+			0 => {
+				PackageMalleability::Malleable
+			},
+			1 => {
+				PackageMalleability::Untractable
+			},
+			_ => return Err(DecodeError::UnknownVersion)
+		};
+		Ok(malleability)
+	}
+}
+
+/// A structure to describe a package content which is generated by ChannelMonitor and
+/// used by OnchainTxHandler to generate and broadcast transactions settling onchain claims.
+///
+/// A package is defined as one or more transactions claiming onchain outputs in reaction
+/// to confirmation of a channel transaction. Those packages might be aggregated to save on
+/// fees, if satisfaction of outputs's witnessScript let's us do so.
+///
+/// As packages are time-sensitive, we fee-bump and rebroadcast them at scheduled intervals.
+/// Failing to confirm a package translate as a loss of funds for the user.
+#[derive(Clone, PartialEq)]
+pub(crate) struct PackageTemplate {
```

Comment:
No need to define structs inside-out order. That's mostly a C-ism. The topmost struct can be defined first in the file. Feel free to ignore, but it's somewhat of pet peeve of mine as it requires reading the file backwards. :)

Reply:
Well I'm coming from a strong C background so quite spontaneous for me to sort structs inside-out :) I'll leave as it is for now but think about it next time I'm defining structs.